text/0049-metric-label-set.md

```diff
@@ -0,0 +1,105 @@
+# Metric `LabelSet` specification
+
+**Status:** `proposed`
+
+Introduce a first-class `LabelSet` API type as a handle on a pre-defined set of labels for the Metrics API.
+
+## Motivation
+
+Labels are the term for key-value pairs used in the OpenTelemetry Metrics API.  Treatment of labels in the Metrics API is especially important for performance across a variety of export strategies.
+
+Label serialization is often one of the most expensive tasks when processing metric events. Creating a `LabelSet` once and re-using it many times can greatly reduce the overall cost of processing many events.
+
+The Metrics API supports three calling conventions: the Handle convention, the Direct convention, and the Batch convention. Each of these conventions stands to benefit when a `LabelSet` is re-used, as it allows the SDK to process the label set once instead of once per call.  Whenever more than one handle will be created with the same labels, more than one instrument will be called directly with the same labels, or more than one batch of metric events will be recorded with the same labels, re-using a `LabelSet` makes it possible for the SDK to improve performance.
+
+## Explanation
+
+Metric instrument APIs which presently take labels in the form `{ Key: Value, ... }` will be updated to take an explicit `LabelSet`.  The `Meter.Labels()` API method supports getting a `LabelSet` from the API, allowing the programmer to acquire a pre-defined label set.  Here are several examples of `LabelSet` re-use.  Assume we have two instruments:
+
+```golang
+var (
+    cumulative = metric.NewFloat64Cumulative("my_counter")
+    gauge      = metric.NewFloat64Gauge("my_gauge")
+)    
+```
+
+Use a `LabelSet` to construct multiple Handles:
+
+```golang
+var (
+    labels  = meter.Labels({ "required_key1": value1, "required_key2": value2 })
```

Comment:
NewLabelSet for the name?

Reply:
I have mixed feelings. I think this is an appropriate name in some languages, but I find it's quite a lot more readable to keep this short. I think `Labels()` is appropriate in Go, at least. I'm also thinking of how an implementation could decide not to allocate anything--for example a streaming SDK could simply record the set of labels and return an offset or an event ID. In that sense it's recording a new set of labels, but it's not returning anything new.

Reply:
Also the spec has a broad declaration at the top stating "As in all OpenTelemetry specifications, these names are examples. Each language committee will decide on the appropriate names based on conventions in that language."

Reply:
Was just for example purpose (to make it clear that a new label set is created). No strong feelings about this.